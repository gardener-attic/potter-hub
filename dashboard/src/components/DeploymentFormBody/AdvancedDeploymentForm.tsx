import * as React from "react";
import AceEditor from "react-ace";

import "ace-builds/src-noconflict/mode-yaml";
import "ace-builds/src-noconflict/theme-xcode";

export interface IAdvancedDeploymentForm {
  appValues?: string;
  handleValuesChange: (value: string) => void;
}

class AdvancedDeploymentForm extends React.Component<IAdvancedDeploymentForm> {
  public render() {
    return (
      <div className="margin-t-normal">
        <AceEditor
          mode="yaml"
          theme="xcode"
          width="100%"
          onChange={this.props.handleValuesChange}
          setOptions={{ showPrintMargin: false }}
          editorProps={{ $blockScrolling: Infinity }}
          value={this.props.appValues}
          className="editor"
        />
        <p>
          <b>Note:</b> Only comments from the original chart values will be preserved.
        </p>
      </div>
    );
  }
}

export default AdvancedDeploymentForm;
