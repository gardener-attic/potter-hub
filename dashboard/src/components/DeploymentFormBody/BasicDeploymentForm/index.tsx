import BasicDeploymentForm from "./BasicDeploymentForm";

export default BasicDeploymentForm;
