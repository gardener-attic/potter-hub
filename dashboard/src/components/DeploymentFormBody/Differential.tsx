import * as jsdiff from "diff";
import { Diff2Html } from "diff2html";
import * as React from "react";

import "diff2html/dist/diff2html.css";

export interface IDifferentialProps {
  title: string;
  oldValues: string;
  newValues: string;
  emptyDiffText: string;
}

class Differential extends React.Component<IDifferentialProps> {
  public render = () => {
    const { oldValues, newValues, title, emptyDiffText } = this.props;
    const sdiff = jsdiff.createPatch(title, oldValues, newValues);
    const outputHtml = Diff2Html.getPrettyHtml(sdiff, {
      inputFormat: "diff",
      showFiles: false,
      matching: "lines",
      maxLineSizeInBlockForComparison: 20,
    });
    return (
      <div className="diff">
        {oldValues === newValues ? (
          <p>{emptyDiffText}</p>
        ) : (
          <div dangerouslySetInnerHTML={{ __html: outputHtml }} />
        )}
      </div>
    );
  };
}

export default Differential;
