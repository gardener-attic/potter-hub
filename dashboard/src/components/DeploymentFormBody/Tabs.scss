.react-tabs {
  &__tab {
    color: black;
    &--selected {
      color: #1598cb;
    }
  }
}

.editor {
  z-index: 0;
}

.diff {
  height: 500px;
  overflow: auto;
}

.diff pre {
  background-color: inherit;
  padding: 0;
  font: 12px / normal "Monaco", "Menlo", "Ubuntu Mono", "Consolas", "source-code-pro", monospace;
  line-height: 12px;
}

.diff table td {
  padding: 0;
}

.diff table td[class*="linenumber"] {
  position: inherit;
}

.diff table td div[class*="code-line"] {
  padding: 0 0.3em;
  min-width: 58em;
}
