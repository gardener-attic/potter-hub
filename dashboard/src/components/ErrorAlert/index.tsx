import ErrorAlertHeader from "./ErrorAlertHeader";
import ErrorSelector from "./ErrorSelector";
import MessageAlert from "./MessageAlert";
import NotFoundErrorAlert from "./NotFoundErrorAlert";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import ServiceBrokersNotFoundAlert from "./ServiceBrokersNotFoundAlert";
import ServiceCatalogNotInstalledAlert from "./ServiceCatalogNotInstalledAlert";
import UnexpectedErrorAlert from "./UnexpectedErrorAlert";

export {
  ErrorAlertHeader,
  ErrorSelector,
  MessageAlert,
  NotFoundErrorAlert,
  PermissionsErrorAlert,
  ServiceBrokersNotFoundAlert,
  ServiceCatalogNotInstalledAlert,
  UnexpectedErrorAlert,
};
