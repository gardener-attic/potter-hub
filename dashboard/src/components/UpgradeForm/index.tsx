import UpgradeForm from "./UpgradeForm";

export default UpgradeForm;
