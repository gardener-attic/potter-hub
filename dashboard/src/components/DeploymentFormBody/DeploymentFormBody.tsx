import { RouterAction } from "connected-react-router";
import * as jsonpatch from "fast-json-patch";
import * as React from "react";
import { AlertOctagon } from "react-feather";
import { Tab, TabList, TabPanel, Tabs } from "react-tabs";

import { definedNamespaces } from "../../shared/Namespace";
import { retrieveBasicFormParams, setValue } from "../../shared/schema";
import { IBasicFormParam, IChartState } from "../../shared/types";
import { getValueFromEvent } from "../../shared/utils";
import ConfirmDialog from "../ConfirmDialog";
import { ErrorSelector } from "../ErrorAlert";
import Hint from "../Hint";
import LoadingWrapper from "../LoadingWrapper";
import AdvancedDeploymentForm from "./AdvancedDeploymentForm";
import BasicDeploymentForm from "./BasicDeploymentForm";

import "react-tabs/style/react-tabs.css";
import Differential from "./Differential";
import "./Tabs.css";

export interface IDeploymentFormBodyProps {
  chartID: string;
  chartVersion: string;
  deployedValues?: string;
  namespace: string;
  releaseVersion?: string;
  selected: IChartState["selected"];
  appValues: string;
  push: (location: string) => RouterAction;
  goBack?: () => RouterAction;
  getChartVersion: (id: string, chartVersion: string) => void;
  setValues: (values: string) => void;
  setValuesModified: () => void;
}

export interface IDeploymentFormBodyState {
  basicFormParameters: IBasicFormParam[];
  restoreDefaultValuesModalIsOpen: boolean;
  modifications?: jsonpatch.Operation[];
}

class DeploymentFormBody extends React.Component<
  IDeploymentFormBodyProps,
  IDeploymentFormBodyState
> {
  public state: IDeploymentFormBodyState = {
    basicFormParameters: [],
    restoreDefaultValuesModalIsOpen: false,
  };

  public componentDidMount() {
    const { chartID, getChartVersion, chartVersion } = this.props;
    getChartVersion(chartID, chartVersion);
  }

  public componentDidUpdate = (prevProps: IDeploymentFormBodyProps) => {
    const { chartID, chartVersion, getChartVersion, selected, appValues } = this.props;

    if (chartVersion !== prevProps.chartVersion) {
      // New version detected
      getChartVersion(chartID, chartVersion);
      return;
    }

    if (
      // If the selected schema changes
      prevProps.selected.schema !== selected.schema ||
      // or the selected values (because it's a different version)
      prevProps.selected.values !== selected.values ||
      // or the current values (and we hadn't process those values yet)
      (prevProps.appValues !== appValues && prevProps.appValues.length === 0)
    ) {
      // Parse the basic parameters
      this.setState({
        basicFormParameters: retrieveBasicFormParams(appValues, selected.schema),
      });
    }
  };

  public render() {
    const { selected, chartID, chartVersion, goBack, namespace } = this.props;
    const { version, versions } = selected;
    if (selected.error) {
      return (
        <ErrorSelector error={selected.error} resource={`Chart "${chartID}" (${chartVersion})`} />
      );
    }
    if (!version || !versions.length) {
      return <LoadingWrapper />;
    }
    return (
      <div>
        <ConfirmDialog
          modalIsOpen={this.state.restoreDefaultValuesModalIsOpen}
          loading={false}
          confirmationText={"Are you sure you want to restore the default chart values?"}
          confirmationButtonText={"Restore"}
          onConfirm={this.restoreDefaultValues}
          closeModal={this.closeRestoreDefaultValuesModal}
        />
        <div>
          <label htmlFor="chartVersion">Version</label>
          <select
            id="chartVersion"
            onChange={this.handleChartVersionChange}
            value={version.attributes.version}
            required={true}
          >
            {versions.map(v => (
              <option key={v.id} value={v.attributes.version}>
                {`Chart: ${v.attributes.version} | App: ${v.attributes.app_version || "-"} `} 
                {this.props.releaseVersion && v.attributes.version === this.props.releaseVersion
                  ? "(current)"
                  : ""}
              </option>
            ))}
          </select>
        </div>
        {this.renderTabs()}
        <div className="margin-t-big">
          <button
            className="button button-primary"
            type="submit"
            disabled={!namespace || namespace === definedNamespaces.all}
          >
            Submit
          </button>
          <button className="button" type="button" onClick={this.openRestoreDefaultValuesModal}>
            Restore Chart Defaults
          </button>
          {goBack && (
            <button className="button" type="button" onClick={goBack}>
              Back
            </button>
          )}
          {(!namespace || namespace === definedNamespaces.all) && (
            <div className="container-tiny" style={{ display: "flex", marginTop: "10px" }}>
              <div className="alert alert-warning AlertBox" role="alert">
                <div className="AlertBoxIconContainer">
                  <AlertOctagon />
                </div>
                Please select a namespace using the namespace selector from the top menu.
              </div>
            </div>
          )}
        </div>
      </div>
    );
  }

  private handleChartVersionChange = (e: React.FormEvent<HTMLSelectElement>) => {
    // TODO(andres): This requires refactoring. Currently, the deploy and upgrade
    // forms behave differently. In the deployment form, a change in the version
    // changes the route but in the case of the upgrade it only changes the state
    const isUpgradeForm = !!this.props.releaseVersion;

    if (isUpgradeForm) {
      const { chartID, getChartVersion } = this.props;
      getChartVersion(chartID, e.currentTarget.value);
    } else {
      this.props.push(
        `/apps/ns/${this.props.namespace}/new/${this.props.chartID}/versions/${e.currentTarget.value}`,
      );
    }
  };

  private handleValuesChange = (value: string) => {
    this.props.setValues(value);
    this.props.setValuesModified();
  };

  private refreshBasicParameters = () => {
    this.setState({
      basicFormParameters: retrieveBasicFormParams(
        this.props.appValues,
        this.props.selected.schema,
      ),
    });
  };

  private renderTabs = () => {
    return (
      <div className="margin-t-normal">
        <Tabs>
          <TabList>
            {this.shouldRenderBasicForm() && (
              <Tab onClick={this.refreshBasicParameters}>
                Form{" "}
                <Hint reactTooltipOpts={{ delayHide: 100 }} id="basicFormHelp">
                  <span>
                    This form has been automatically generated based on the chart schema.
                    <br />
                    This feature is currently in a beta state. If you find an issue please report it{" "}
                    <a
                      target="_blank"
                      rel="noopener noreferrer"
                      href="https://github.com/gardener/potter-hub/issues/new/choose"
                    >
                      here.
                    </a>
                  </span>
                </Hint>
              </Tab>
            )}
            <Tab>Values (YAML)</Tab>
            <Tab>Changes</Tab>
          </TabList>
          {this.shouldRenderBasicForm() && (
            <TabPanel>
              <BasicDeploymentForm
                params={this.state.basicFormParameters}
                handleBasicFormParamChange={this.handleBasicFormParamChange}
                appValues={this.props.appValues}
                handleValuesChange={this.handleValuesChange}
              />
            </TabPanel>
          )}
          <TabPanel>
            <AdvancedDeploymentForm
              appValues={this.props.appValues}
              handleValuesChange={this.handleValuesChange}
            />
          </TabPanel>
          <TabPanel>{this.renderDiff()}</TabPanel>
        </Tabs>
      </div>
    );
  };

  private handleBasicFormParamChange = (param: IBasicFormParam) => {
    return (e: React.FormEvent<any>) => {
      this.props.setValuesModified();
      const value = getValueFromEvent(e);
      this.setState({
        // Change param definition
        basicFormParameters: this.state.basicFormParameters.map(p =>
          p.path === param.path ? { ...param, value } : p,
        ),
      });
      // Change raw values
      this.props.setValues(setValue(this.props.appValues, param.path, value));
    };
  };

  // The basic form should be rendered if there are params to show
  private shouldRenderBasicForm = () => {
    return Object.keys(this.state.basicFormParameters).length > 0;
  };

  private closeRestoreDefaultValuesModal = () => {
    this.setState({ restoreDefaultValuesModalIsOpen: false });
  };

  private openRestoreDefaultValuesModal = () => {
    this.setState({ restoreDefaultValuesModalIsOpen: true });
  };

  private restoreDefaultValues = () => {
    if (this.props.selected.values) {
      this.props.setValues(this.props.selected.values);
      this.setState({
        basicFormParameters: retrieveBasicFormParams(
          this.props.selected.values,
          this.props.selected.schema,
        ),
      });
    }
    this.setState({ restoreDefaultValuesModalIsOpen: false });
  };

  private renderDiff = () => {
    let oldValues = "";
    let title = "";
    let emptyDiffText = "";
    if (this.props.deployedValues) {
      // If there are already some deployed values (upgrade scenario)
      // We compare the values from the old release and the new one
      oldValues = this.props.deployedValues;
      title = "Difference from deployed version";
      emptyDiffText = "The values for the new release are identical to the deployed version.";
    } else {
      // If it's a new deployment, we show the different from the default
      // values for the selected version
      oldValues = this.props.selected.values || "";
      title = "Difference from chart defaults";
      emptyDiffText = "No changes detected from chart defaults.";
    }
    return (
      <Differential
        title={title}
        oldValues={oldValues}
        newValues={this.props.appValues}
        emptyDiffText={emptyDiffText}
      />
    );
  };
}

export default DeploymentFormBody;
