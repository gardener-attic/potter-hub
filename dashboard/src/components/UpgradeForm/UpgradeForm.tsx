import { RouterAction } from "connected-react-router";
import * as jsonpatch from "fast-json-patch";
import { JSONSchema4 } from "json-schema";
import * as React from "react";
import * as YAML from "yaml";

import { deleteValue, setValue } from "../../shared/schema";
import { IChartState, IChartVersion } from "../../shared/types";
import DeploymentFormBody from "../DeploymentFormBody/DeploymentFormBody";
import { ErrorSelector } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";

export interface IUpgradeFormProps {
  appCurrentVersion: string;
  appCurrentValues?: string;
  chartName: string;
  namespace: string;
  releaseName: string;
  repo: string;
  error: Error | undefined;
  selected: IChartState["selected"];
  deployed: IChartState["deployed"];
  upgradeApp: (
    version: IChartVersion,
    releaseName: string,
    namespace: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
  goBack: () => RouterAction;
  fetchChartVersions: (id: string) => Promise<IChartVersion[] | undefined>;
  getChartVersion: (id: string, chartVersion: string) => void;
}

interface IUpgradeFormState {
  appValues: string;
  valuesModified: boolean;
  isDeploying: boolean;
  modifications?: jsonpatch.Operation[];
  deployedValues?: string;
}

class UpgradeForm extends React.Component<IUpgradeFormProps, IUpgradeFormState> {
  public state: IUpgradeFormState = {
    appValues: this.props.appCurrentValues || "",
    isDeploying: false,
    valuesModified: false,
  };

  public componentDidMount() {
    const chartID = `${this.props.repo}/${this.props.chartName}`;
    this.props.fetchChartVersions(chartID);
  }

  public componentDidUpdate = (prevProps: IUpgradeFormProps) => {
    let modifications = this.state.modifications;
    // applyModifications is an expensive operatior, that's why it's only defined within
    // the componentDidUpdate scope
    const applyModifications = (mods: jsonpatch.Operation[], appValues: string) => {
      // And we add any possible change made to the original version
      if (mods.length) {
        mods.forEach(modification => {
          if (modification.op === "remove") {
            appValues = deleteValue(appValues, modification.path);
          } else {
            // Transform the modification as a ReplaceOperation to read its value
            const value = (modification as jsonpatch.ReplaceOperation<any>).value;
            appValues = setValue(appValues, modification.path, value);
          }
        });
      }
      return appValues;
    };

    if (this.props.deployed.values && !modifications) {
      // Calculate modifications from the default values
      const defaultValuesObj = YAML.parse(this.props.deployed.values);
      const deployedValuesObj = YAML.parse(this.props.appCurrentValues || "");
      modifications = jsonpatch.compare(defaultValuesObj, deployedValuesObj);
      const values = applyModifications(modifications, this.props.deployed.values);
      this.setState({ modifications });
      this.setState({ appValues: values, deployedValues: values });
    }

    if (prevProps.selected.version !== this.props.selected.version && !this.state.valuesModified) {
      // Apply modifications to the new selected version
      const appValues = modifications
        ? applyModifications(modifications, this.props.selected.values || "")
        : this.props.selected.values || "";
      this.setState({ appValues });
    }
  };

  public render() {
    const { namespace, releaseName, error, selected } = this.props;
    if (error) {
      return (
        <ErrorSelector error={error} namespace={namespace} action="update" resource={releaseName} />
      );
    }
    if (selected.versions.length === 0) {
      return <LoadingWrapper />;
    }
    const chartID = `${this.props.repo}/${this.props.chartName}`;
    return (
      <form className="container padding-b-bigger" onSubmit={this.handleDeploy}>
        <div className="row">
          <div className="col-12">
            <h2>{`${this.props.releaseName} (${chartID})`}</h2>
          </div>
          <div className="col-8">
            <DeploymentFormBody
              chartID={chartID}
              chartVersion={this.props.appCurrentVersion}
              deployedValues={this.state.deployedValues}
              namespace={this.props.namespace}
              releaseVersion={this.props.appCurrentVersion}
              selected={this.props.selected}
              push={this.props.push}
              goBack={this.props.goBack}
              getChartVersion={this.props.getChartVersion}
              setValues={this.handleValuesChange}
              appValues={this.state.appValues}
              setValuesModified={this.setValuesModified}
            />
          </div>
        </div>
      </form>
    );
  }

  public setValuesModified = () => {
    this.setState({ valuesModified: true });
  };

  public handleValuesChange = (value: string) => {
    this.setState({ appValues: value });
  };

  public handleDeploy = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { selected, push, upgradeApp, releaseName, namespace } = this.props;
    const { appValues } = this.state;

    this.setState({ isDeploying: true });
    if (selected.version) {
      const deployed = await upgradeApp(
        selected.version,
        releaseName,
        namespace,
        appValues,
        selected.schema,
      );
      this.setState({ isDeploying: false });
      if (deployed) {
        push(`/apps/ns/${namespace}/${releaseName}`);
      }
    }
  };
}

export default UpgradeForm;
