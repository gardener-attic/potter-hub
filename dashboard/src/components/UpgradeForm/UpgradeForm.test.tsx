import { mount, shallow } from "enzyme";
import * as React from "react";
import itBehavesLike from "../../shared/specs";

import { IChartState, IChartVersion, UnprocessableEntity } from "../../shared/types";
import DeploymentFormBody from "../DeploymentFormBody/DeploymentFormBody";
import { ErrorSelector } from "../ErrorAlert";
import UpgradeForm, { IUpgradeFormProps } from "./UpgradeForm";

const versions = [{ id: "foo", attributes: { version: "1.2.3" } }] as IChartVersion[];

const defaultProps = {
  appCurrentVersion: "1.0.0",
  appCurrentValues: "foo: bar",
  chartName: "my-chart",
  namespace: "default",
  releaseName: "my-release",
  repo: "my-repo",
  selected: { versions } as IChartState["selected"],
  deployed: {} as IChartState["deployed"],
  upgradeApp: jest.fn(),
  push: jest.fn(),
  goBack: jest.fn(),
  fetchChartVersions: jest.fn(),
  getChartVersion: jest.fn(),
  error: undefined,
} as IUpgradeFormProps;

itBehavesLike("aLoadingComponent", {
  component: UpgradeForm,
  props: { ...defaultProps, selected: { versions: [] } },
});

it("fetches the available versions", () => {
  const fetchChartVersions = jest.fn();
  shallow(<UpgradeForm {...defaultProps} fetchChartVersions={fetchChartVersions} />);
  expect(fetchChartVersions).toHaveBeenCalledWith(`${defaultProps.repo}/${defaultProps.chartName}`);
});

describe("renders an error", () => {
  it("renders a custom error if the deployment failed", () => {
    const wrapper = shallow(
      <UpgradeForm
        {...defaultProps}
        selected={
          {
            version: { attributes: {} },
            versions: [{ id: "foo", attributes: {} }],
          } as IChartState["selected"]
        }
        error={new UnprocessableEntity("wrong format!")}
      />,
    );
    wrapper.setState({ latestSubmittedReleaseName: "my-app" });
    expect(wrapper.find(ErrorSelector).exists()).toBe(true);
    expect(wrapper.find(ErrorSelector).html()).toContain(
      "Sorry! Something went wrong processing my-release",
    );
    expect(wrapper.find(ErrorSelector).html()).toContain("wrong format!");
  });
});

it("renders the full UpgradeForm", () => {
  const wrapper = shallow(
    <UpgradeForm {...defaultProps} selected={{ versions, version: versions[0] }} />,
  );
  expect(wrapper).toMatchSnapshot();
});

it("defaults the upgrade version to the current version", () => {
  // helm upgrade is the only way to update the values.yaml, so upgrade is
  // often used by users to update values only, so we can't default to the
  // latest version on the assumption that they always want to upgrade.
  const wrapper = shallow(
    <UpgradeForm {...defaultProps} selected={{ versions, version: versions[0] }} />,
  );

  expect(wrapper.find(DeploymentFormBody).prop("chartVersion")).toBe("1.0.0");
});

it("forwards the appValues when modified", () => {
  const wrapper = shallow(<UpgradeForm {...defaultProps} />);
  const handleValuesChange: (v: string) => void = wrapper
    .find(DeploymentFormBody)
    .prop("setValues");
  handleValuesChange("foo: bar");

  expect(wrapper.state("appValues")).toBe("foo: bar");
  expect(wrapper.find(DeploymentFormBody).prop("appValues")).toBe("foo: bar");
});

it("triggers an upgrade when submitting the form", done => {
  const releaseName = "my-release";
  const namespace = "default";
  const appValues = "foo: bar";
  const schema = { properties: { foo: { type: "string" } } };
  const upgradeApp = jest.fn(() => true);
  const push = jest.fn();
  const wrapper = mount(
    <UpgradeForm
      {...defaultProps}
      selected={{ versions, version: versions[0], schema }}
      upgradeApp={upgradeApp}
      push={push}
      namespace={namespace}
    />,
  );
  wrapper.setState({ releaseName, appValues });
  wrapper.find("form").simulate("submit");
  expect(upgradeApp).toHaveBeenCalledWith(versions[0], releaseName, namespace, appValues, schema);
  setTimeout(() => {
    expect(push).toHaveBeenCalledWith("/apps/ns/default/my-release");
    done();
  }, 1);
});

describe("when receiving new props", () => {
  it("should calculate the modifications from the default and the current values", () => {
    const currentValues = "a: b\nc: d\n";
    const defaultValues = "a: b\n";
    const expectedModifications = [{ op: "add", path: "/c", value: "d" }];
    const wrapper = shallow(<UpgradeForm {...defaultProps} appCurrentValues={currentValues} />);
    wrapper.setProps({ deployed: { values: defaultValues } });

    expect(wrapper.state("modifications")).toEqual(expectedModifications);
    expect(wrapper.state("appValues")).toEqual(currentValues);
  });

  it("should apply modifications if a new version is selected", () => {
    const defaultValues = "a: b\n";
    const modifications = [{ op: "add", path: "/c", value: "d" }];
    const wrapper = shallow(<UpgradeForm {...defaultProps} />);
    wrapper.setState({ modifications });
    wrapper.setProps({ selected: { versions: [], version: {}, values: defaultValues } });

    expect(wrapper.state("appValues")).toEqual("a: b\nc: d\n");
  });

  it("won't apply changes if the values have been manually modified", () => {
    const userValues = "a: b\n";
    const modifications = [{ op: "add", path: "/c", value: "d" }];
    const wrapper = shallow(<UpgradeForm {...defaultProps} />);
    wrapper.setState({ modifications, valuesModified: true, appValues: userValues });
    wrapper.setProps({ selected: { versions: [], version: {} } });

    expect(wrapper.state("appValues")).toEqual(userValues);
  });

  [
    {
      description: "should merge modifications from the values and the new version defaults",
      defaultValues: "foo: bar\n",
      deployedValues: "foo: bar\nmy: var\n",
      newDefaultValues: "notFoo: bar",
      result: "notFoo: bar\nmy: var\n",
    },
    {
      description: "should modify the default values",
      defaultValues: "foo: bar\n",
      deployedValues: "foo: BAR\nmy: var\n",
      newDefaultValues: "foo: bar",
      result: "foo: BAR\nmy: var\n",
    },
    {
      description: "should delete an element in the defaults",
      defaultValues: "foo: bar\n",
      deployedValues: "my: var\n",
      newDefaultValues: "foo: bar\n",
      result: "my: var\n",
    },
    {
      description: "should add an element in an array",
      defaultValues: `foo:
  - foo1:
    bar1: value1
`,
      deployedValues: `foo:
  - foo1: 
    bar1: value1
  - foo2: 
    bar2: value2
`,
      newDefaultValues: `foo:
    - foo1:
      bar1: value1
`,
      result: `foo:
  - foo1: 
    bar1: value1
  - foo2: 
    bar2: value2
`,
    },
    {
      description: "should delete an element in an array",
      defaultValues: `foo:
  - foo1:
    bar1: value1
  - foo2:
    bar2: value2
`,
      deployedValues: `foo:
  - foo1: 
    bar1: value1
`,
      newDefaultValues: `foo:
  - foo1:
    bar1: value1
  - foo2:
    bar2: value2
`,
      result: `foo:
  - foo1: 
    bar1: value1
`,
    },
    {
      description: "set a value with dots and slashes in the key",
      defaultValues: "foo.bar/foobar: ",
      deployedValues: "foo.bar/foobar: value",
      newDefaultValues: "foo.bar/foobar: ",
      result: "foo.bar/foobar: value\n",
    },
  ].forEach(t => {
    it(t.description, () => {
      const deployed = {
        values: t.defaultValues,
        requested: true,
      };
      const newSelected = {
        ...defaultProps.selected,
        version: { trigger: "change" },
        values: t.newDefaultValues,
      };
      const wrapper = shallow(
        <UpgradeForm {...defaultProps} appCurrentValues={t.deployedValues} />,
      );
      wrapper.setProps({ deployed });

      // Apply new version
      wrapper.setProps({ selected: newSelected });
      expect(wrapper.state("appValues")).toEqual(t.result);
    });
  });
});

it("shows, by default, the default values of the deployed chart plus any modification", () => {
  const wrapper = shallow(<UpgradeForm {...defaultProps} appCurrentValues="foo: not-bar" />);
  wrapper.setProps({ deployed: { values: "# A comment\nfoo: bar\n" } as IChartState["deployed"] });
  const expectedValues = "# A comment\nfoo: not-bar\n";
  expect(wrapper.find(DeploymentFormBody).prop("deployedValues")).toBe(expectedValues);
});
