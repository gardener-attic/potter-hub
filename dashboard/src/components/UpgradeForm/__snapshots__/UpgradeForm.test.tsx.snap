// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;

exports[`renders the full UpgradeForm 1`] = `
<form
  className="container padding-b-bigger"
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <h2>
        my-release (my-repo/my-chart)
      </h2>
    </div>
    <div
      className="col-8"
    >
      <DeploymentFormBody
        appValues="foo: bar"
        chartID="my-repo/my-chart"
        chartVersion="1.0.0"
        getChartVersion={[MockFunction]}
        goBack={[MockFunction]}
        namespace="default"
        push={[MockFunction]}
        releaseVersion="1.0.0"
        selected={
          Object {
            "version": Object {
              "attributes": Object {
                "version": "1.2.3",
              },
              "id": "foo",
            },
            "versions": Array [
              Object {
                "attributes": Object {
                  "version": "1.2.3",
                },
                "id": "foo",
              },
            ],
          }
        }
        setValues={[Function]}
        setValuesModified={[Function]}
      />
    </div>
  </div>
</form>
`;
